docs/attributes-registry/pipeline.md

```diff
@@ -0,0 +1,22 @@
+<!--- Hugo front matter used to generate the website version of this page:
+--->
+
+<!-- NOTE: THIS FILE IS AUTOGENERATED. DO NOT EDIT BY HAND. -->
+<!-- see templates/registry/markdown/attribute_namespace.md.j2 -->
+
+# Pipeline
+
+## Pipeline Attributes
+
+This group describes attributes specific to pipelines.
```

Comment:
Given that there are so many [different terminologies](https://github.com/cdfoundation/sig-interoperability/blob/main/docs/tools-terminology.md) used by different tools, we should make sure that pipeline and step are well defined. I found the [CDEvents terminology](https://github.com/cdfoundation/sig-interoperability/blob/main/docs/tools-terminology.md#cdevents) particularly compelling.

It basically comes down to there being a hierarchy grouping related parts together and distinguishing between the definition of pipelines/steps and their instantiation/execution. Some tools have additional entities to help group together pipelines or steps (folders, groups, stages).

For example for illustration:
```mermaid
flowchart LR
    Group -->|contains| Pipeline
    Pipeline -->|contains| Stage
    Stage -->|contains| Task
    PipelineRun -->|is instance of| Pipeline
    PipelineRun -->|contains| TaskRun
    TaskRun -->|is instance of| Task
```

I think this is also what @pablochacin's comment pointed out: https://github.com/open-telemetry/semantic-conventions/issues/915#issuecomment-2083210096

Reply:
I am not sure, I understand your proposal. OpenTelemetry does not store definitions of the object, it stores the runtime data of the execution of your CI/CD to make an analysis of the data, so the concept of `PipelineRun` in CDEvents is mapped 1:1 as `pipeline` and the concept of `TaskRun` as a `step` not as instances of object definitions.

Reply:
What I'm proposing is to store a reference to `Pipeline` (eg id, name, url) so that it's possible to filter for related pipeline runs. Without this we would have this filtering in either vendor specific attributes or by having it encoded as part of the `pipeline.id`

Reply:
Do you mean to relate the build and the pipeline with a field? AFAIU, the build object stores will have pipeline data related. The objects on Tpentelemetry are more NO-SQL objects than relational database objects. So a build will be something like this:

```json
{
  "build": {
    "id": "bar-id",
    "name": "bar-name",
    ...
  },
  "pipeline":{
    "id": "foo-id",
    "name": "foo-name",
    ...
  }
}
```

You have a relation between the pipeline and the build in the same object. There are no nested objects.

Reply:
> The objects on Opentelemetry are more NO-SQL objects than relational database objects.

I agree. That's why we would need to include all relevant attributes if we want to filter/query by them.

My questions are more:
* Which of these attributes make sense to define in CICD SemConv?
* How should they be defined (given the different terminologies in use by tools)?

> So a build will be something like this:

Would build.id/name correspond to the CDEvents definition of a PipelineRun and pipeline.id/name to Pipeline?

Reply:
@christophe-kamphaus-jemmic  - I don't think so. I think `pipeline` correlates to `pipelineRun`. The reason I say this is because CDEvents are in the context of an event. Whereas these attributes are common and signal trancending. I think once we get into the realm of actually defining conventions for each signal (should they differ) and saying `x, y, z` signal supports extending to CDEvents or Eiffel, then we can start to build out a translator (perhaps) that maps `x` attribute in OTEL to `y` attribute in another spec. 

I think there's going to have to be a mapping/translator at some point, but I'm also hoping that tools start to adopt the semantic conventions as much as possible. Additionally, I think that defining how vendors can easily extend with their own vendor specific attributes will play a critical role in both adoption and interoperability. I believe that was discussed a bit on the last SemConv meeting & I haven't got to pair on that yet. Nor have I paired on the db enum like semantics that could also be added here. Starting with small iterations and fast feedback to get this exact type of discussion going!

Reply:
> The reason I say this is because CDEvents are in the context of an event.

Which event context do you mean?

1. The event which triggers a pipeline run. CICD is inherently event-based. An event (timer, SCM event, user interaction, webhook/MQ event, …) triggers CICD actions (pipeline, GitOps sync, …).

2. CDEvents are transmitted as events. However the entities that these events refer to (eg. Pipeline, PipelineRun) are independent from how they are transferred be it event or span. As such any CICD attributes we define in Otel SemConv can refer to these same entities.
For example: span / traces of a pipeline run could have the attributes:
```
cicd.pipeline.id=foo
cicd.pipelineRun.id=foo/7
```
and a span of a TaskRun (child span of the above span) could have the attributes:
```
cicd.pipeline.id=foo
cicd.pipelineRun.id=foo/7
cicd.stage.name=Build
cicd.task.name=sh
```
> Whereas these attributes are common and signal trancending.

I'm not sure I get what you are getting at. Let me try to phrase it as I see it:
The attributes defined in Otel SemConv should be as generic as possible and defined well enough to be easily translatable to CDEvents or Eiffel and implemented by tool vendors without causing confusion due to differing interpretations (eg one tool putting the Pipeline Id in `pipeline.id` while another puts the PipelineRun Id in `pipeline.id`). Ie OTel should help the different tools to be interoperable.

> I think there's going to have to be a mapping/translator at some point ...

Agreed, looking forward to it.

Reply:
> Which event context do you mean?

As in the context of a Signal (Log) where an Event is a LogRecord. I simply meant that the specification was built around everything being emitted as an event. At least that's the third party perspective I have when reading through the specification. After discussions in general I know those folks have done a lot of thinking in-general about interoperability.

> CDEvents are transmitted as events. However the entities that these events refer to (eg. Pipeline, PipelineRun) are independent from how they are transferred be it event or span.

That was my point. CDEvents are transmitted as events and inherently thought about in the context of Events. When I read it, it's the difference between saying:

> This is my `pipeline`

vs.

> This was my `pipelineRun`

If we want to say `pipeline` maps to `pipelineRun` this mapping would occur in the convention for signal. 

I simply think this mapping isn't suited for this pull request adding things to the registry and would be best suited for a separate issue / signal specific pull request.

Reply:
At this point I'm not sure what is been proposed regarding this two comments:

@christophe-kamphaus-jemmic 
> The attributes defined in Otel SemConv should be as generic as possible and defined well enough to be easily translatable to CDEvents or Eiffel and implemented by tool vendors without causing confusion due to differing interpretations (eg one tool putting the Pipeline Id in pipeline.id while another puts the PipelineRun Id in pipeline.id). Ie OTel should help the different tools to be interoperable.

@adrielp 
> If we want to say pipeline maps to pipelineRun this mapping would occur in the convention for signal.

Are we arguing not to make a difference between `pipeline` and `pipelineRun` and leave to the tool the interpretation of a `pipeline` attribute?

I don't think mapping `pipeline` to `pipelineRun` is something that makes sense in general. A `pipeline` (or workflow) is a definition or template and a `pipelineRun` is an execution or instance of this template. I think this is common in most systems, if not all I've worked with. 

The relationship is 1-to-many and we can't assume the `pipeline` can be inferred from the `pipelineRun` in general. Probably some systems use a convention where `pipelineRun` has the form `pileline/<runid>` but I'm not sure we should rely on this. 

Moreover, letting the tools the interpretation of what a `pipeline` is (a definition of a template or an instance) introduces more ambiguity IMO.

Reply:
I'm arguing that we should make a difference between `pipeline` and `pipelineRun` for exactly the reasons that @pablochacin has listed. Whether we call them `pipeline` and `pipelineRun` or something else we can discuss, but they should be separate and well defined to prevent ambiguity from differing tool interpretations.

A similar argument can be made for Task and TaskRun: the `pipeline.step` defined in this document seem to refer to the TaskRun. Should there also be a reference to the definition of the step? In any case the definition of `pipeline.step` should make it clear whether we refer to the definition or the instance.

The other entities with 1-to-many relationships in my original comment (group, stage) help deal with the organization of pipelines and tasks and could be added as optional later. They help cope with the complexity of many pipelines or long pipelines.

> I simply meant that the specification was built around everything being emitted as an event.

I agree that the CDEvents specs is focused on events. However the terminology that I picked up from the CDFoundation [tools-terminology.md#cdevents](https://github.com/cdfoundation/sig-interoperability/blob/main/docs/tools-terminology.md#cdevents) page like Pipeline and PipelineRun seem to me like they could be used in spans as well since the concepts remain the same no matter how data referring to them is transmitted.

@adrielp Let's discuss this in Monday's SemConv meeting. I have a feeling that we are starting to argue in circles.

Reply:
@christophe-kamphaus-jemmic - we're all good. this is a nice conversation fleshing out different points and clarifications. With all the questions that have occurred, I've deduced that I need to provide in depth descriptions and references for the attributes. Will be working on that this week & still have other pieces of feedback to address and contemplate. I will try to be there today but I have a conflict so it may only be the last 15 minutes of the call.

Reply:
The current pipeline definition and attribute descriptions in https://github.com/open-telemetry/semantic-conventions/blob/1ea5c57c2a3f40719de7f426d38b094b984c2dc9/docs/attributes-registry/cicd.md define pipelines well enough. For runs the description could include that it is an instance / execution of a pipeline or task respectively.
Group/stages can be defined as vendor specific extensions as part of https://github.com/open-telemetry/semantic-conventions/issues/1193